diff --git a/article/creating-a-blog-article-reviewer-with-ai.mdx b/article/creating-a-blog-article-reviewer-with-ai.mdx
new file mode 100644
index 0000000..d6fd244
--- /dev/null
+++ b/article/creating-a-blog-article-reviewer-with-ai.mdx
@@ -0,0 +1,30 @@
+---
+layout: post
+title: "Who Needs Humans Code Reviews When We Have AI?"
+excerpt: "We automated proofreading of our blog articles with a GitHub action leveraging an LLM. It even reviewed this very article!"
+cover_image: "./images/large.webp"
+thumbnail_image: "./images/small.webp"
+authors:
+  - thiery
+tags:
+  - ai
+  - bot
+  - github
+---
+
+In our quest to master AI and automation, we have been experimenting with pull requests reviews. It turns out this very blog post was written in a code editor and submitted as a pull request. How hard would it be to have an AI review the pull request?
+
+We built a bot leveraging an LLM from OpenAI that reviews the pull requests on the blog repository. The bot provides comments and suggestions to improve the article. The bot is integrated with GitHub actions, so it runs automatically on every pull request. It even reviewed this very article!
+
+![An AI comment](images/ai-comment.png)
+
+## The Big Picture
+
+I built a [GitHub action](https://docs.github.com/en/actions) that triggers on every pull request. The action retrieves the diff of the pull request, sends it to [OpenAI's chat completion API](https://platform.openai.com/docs/guides/text-generation) with custom instructions, and uses the result to create a review on the pull request with the comments and suggestions.
+
+Calling the OpenAI API is pretty straightforward, thanks to the [openai package](https://github.com/openai/openai-node). The difficulties are elsewhere:
+
+- How to properly prompt the LLM to review an article?
+- How to use the LLM response to create a review with the GitHub API?
+
+![The Quest for the Right Prompt](./images/theQuestForTheRightPrompt.webp)
